File: 프로그래머스/2/76502. 괄호 회전하기/괄호 회전하기.cpp
#include <string>
#include <vector>

using namespace std;

int solution(string s) {
    int answer = 0;
        for (int i = 0; i < s.size(); i++)
    {
        vector<char> temp;
        bool chk = true;
        for (int j = 0; j < s.size(); j++)
        {
            char input = s[(i + j) % s.size()];
            if (input == '{' || input == '(' || input == '[')
            {
                temp.push_back(input);
            }
            else
            {
                if (temp.empty())
                {
                    chk = false;
                    break;
                }

                if (input == '}')
                {

                    if(temp.back() == '{')                
                        temp.pop_back();
                    else
                    {
                        chk = false;
                        break;
                    }

                }

                if (input == ')')
                {

                    if (temp.back() == '(')
                        temp.pop_back();
                    else
                    {
                        chk = false;
                        break;
                    }

                }
                if (input == ']')
                {

                    if (temp.back() == '[')
                        temp.pop_back();
                    else
                    {
                        chk = false;
                        break;
                    }

                }
            }
        }
        if (temp.empty() && chk) answer++;
    }
    return answer;
}

File: 프로그래머스/2/76502. 괄호 회전하기/괄호 회전하기.cs
using System;
using System.Linq;
using System.Collections.Generic;
public class Solution {
    public int solution(string s) {
        int answer = 0;
        var queue = new Queue<char>();
var stack = new Stack<char>();

foreach(var item in s)
{
    queue.Enqueue(item);
}

for(int i = 0; i< s.Length;i++)
{
    stack.Clear();
    bool chk = true;
    foreach (var item in queue)
    {
        if (stack.Count == 0)
        {
            if (item == ']' || item == '}' || item == ')')
            {
                chk = false;
                break;
            }
            else stack.Push(item);
        }
        else
        {
            if(item == '[' || item == '{' || item == '(')
            {
                stack.Push(item);
            }
            else
            {
                if (item == ']' && stack.Peek() == '[')
                    stack.Pop();
                else if (item == '}' && stack.Peek() == '{')
                    stack.Pop();
                else if (item == ')' && stack.Peek() == '(')
                    stack.Pop();
                else
                {
                    chk = false;
                    break;
                }
            }
        }
    }
    if(stack.Count == 0 && chk)
        answer++;

    var go = queue.Dequeue();
    queue.Enqueue(go);
}
        return answer;
    }
}